### DISCUSSION TITLE
Using Insiders Blog plugin without defined Navigation

### DISCUSSION THREAD

**Discussion by polymorcodeus** [squidfunk/mkdocs-material#5132] [Q&A] [answered]
Heya-

I'm trying to figure out if it's possible to enable the Blog insiders plugin without a defined nav section. We have a main site with a number of remote repos "publishing" into our main site, to keep the docs close to the tooling so we just rely on a non-defined `nav` section.

However it looks? like this might be incompatible with the Blog plugin as I get this error when trying to build:
`ERROR    -  Blog root 'blog/index.md' not in navigation.`

I'm currently pulling down the latest Insiders version - just in case - however the documentation definitely seems to indicate you have to hard define it in the nav section. Though I'm really hoping I'm just missing something obvious.

**Comment by squidfunk** [comment]
Currently not possible, yes. We could in theory add the blog "somewhere", but that's likely not the location you would imagine. I'm not sure that this is a good idea. Have you considered using the [`awesome-pages`](https://github.com/lukasgeiter/mkdocs-awesome-pages-plugin) plugin, which allows you to partially define your navigation? It might also be possible to just scope this to the blog section. Not sure though, I'm not a user.

**Reply by polymorcodeus** [reply]
I tried adding the blog to the nav tabs to add it somewhere, but it still didn't like that.

I'll have a look at `awesome-pages`. Worst case I'll do some dynamic nav creation at publish time.

**Reply by squidfunk** [reply]
I'll investigate whether it's possible to "add it somewhere" if and only if the `nav` is not explicitly defined. If the `nav` is defined in `mkdocs.yml`, it's likely the best idea to keep it mandatory. However, I'm not sure whether this might interfere with some other plugins like the aforementioned `awesome-pages` plugin.

**Reply by polymorcodeus** [reply]
💯 Thanks! And all good. 

I'll check out the plugin in the meantime. One way or another, we'll get our blog added.

**Reply by squidfunk** [reply]
Please share your experience with the plugin or how you reach a solution in the end – that would be valuable feedback!

**Reply by squidfunk** [reply]
I've investigated and think that allowing empty navigation would make things much more complex. The reason is that we must generate all computed pages (i.e. blog index pages, as well as archive and category indexes) and make them visible to MkDocs _before_ the `on_nav` event, or MkDocs will complain about pages that are not in the navigation. For this to work properly, we must jump through some hoops, because MkDocs isn't really built for auto-generated pages.

Implementing an exception for when no navigation is specified would mean we would need to add another edge case where this logic is not applied. I'm not sure if this is of benefit to many users, so I think our time is better invested in other features. From what I understand reading the code of the `awesome-pages` plugin, it looks like it does its "magic" inside the `on_config` hook, which runs before the `on_files` hook used in the blog plugin where the navigation is checked.

I believe that this plugin should solve the issue at hand.

**Reply by polymorcodeus** [reply]
I would love to say I got this working ... sadly no go. Even though it should be doing its magic first, it definitely doesn't appear to be. My testing is based off of this [closed issue](https://github.com/lukasgeiter/mkdocs-awesome-pages-plugin/issues/40). I've also done a bunch of testing trying to leverage the `.pages` files, but matching inside folders doesn't work as advertised, though still can't get past the blog/index.md not found anyways so left that abandoned.

Test Site looks like this:
```
.
├── docs
│   ├── blog
│   │   ├── index.md
│   │   └── posts
│   │       ├── hello-world.md
│   │       └── hello-world23.md
│   ├── hello-world.md
│   ├── index.md
│   └── not-blog
│       ├── hello-world.md
│       └── index.md
└── mkdocs.yml
```

Nav section in `mkdocs.yml`
```
nav:
  - Blog:
    - blog/index.md
  - ...
```
Without the `blog` plugin you get:
![image](https://user-images.githubusercontent.com/11621557/222862777-aa46d07f-19f9-45b6-94e3-656268098a78.png)
as expected ... going to have some problems once the "Blog" plugin kicks in with Archive/Categories folders, but ignoring a split-blog scenario for now.

Once I add the blog plugin (after OR before awesome-pages), same error as before: `ERROR    -  Blog root 'blog/index.md' not in navigation.` 

Ideally, I would have this as the nav section in `mkdocs.yml` so that it would capture all of the files in blog but wanted to explicitly define the `blog/index.md` to try and get around the error.
```
nav:
  - ... | regex=^[blog]
  - ...
```

Unless I've grossly missed something, which I won't count on, next step is just going to be generating the nav section at publish time in CI and just get a list of all the folders inside `/docs` and populate the nav section from there.

**Reply by squidfunk** [reply]
Thanks for testing. So the blog plugin is currently reserved to sponsors, but when the next funding goal is hit it will be released to the public. Making it available to the public, I'd expect increasing adoption, and we might reconsider supporting additional configuration methodologies then as we learn about them. Additionally, with the source being available, other users can start working on PRs adding this functionality. However currently, this is not a priority. Thanks for your understanding!

**Reply by polymorcodeus** [reply]
No worries at all! Dynamically generating the nav will just get me going. After that might have a look at the plugin and seeing if there is anything to do at that level on the long term. 

It looks like it should have fit the bill perfectly - so it might just need a little tweaking, who knows! Thanks for the suggestion!

**Reply by polymorcodeus** [reply]
Got some time with this - figured I'd add my findings to date.

So it looks like after the `on_config` portion, it actually removes the nav config to prevent the console output of "missing files", then re-adds it in the `on_nav` step. So I just need to add it to the side then merge it with the nav portion that get's autogenerated with the blog plugin and then I think it SHOULD work after that. If I get something less hacky then my current stay, I'll post the relevant patches to the plugin so others could benefit.

**Reply by polymorcodeus** [reply]
FULL disclosure - this is incredibly hacky and I'm sure this could be worked into something slightly more production worthy BUT this works for my narrow use case. I've briefly tested this for multiple rest sections, YMMV. 

Also, I'm not remotely a python dev, so apologies in advance. 

TL;DR - stores `REST_PLACE_HOLDER` elements and removes them in `config` stage as needed. Uses updated `get_nav` function that ignores the fact that a `config["nav"]` is defined and inserts them back into the nave.items list. You'll get a `INFO:    The following pages exist in the docs directory, but are not included in the "nav" configuration:` that I haven't bothered to try and supress.

NAV section:
```
nav:
  - index.md
  - Blog:
    - blog/index.md
  - tags.md
  - ...
  ```

Personal Fork of mkdocs-awesome-pages-plugin:
```
plugin.py - replaced func(s) on_nav and on_config with below
...
    def on_nav(self, nav: MkDocsNavigation, config: Config, files: Files):
        explicit_nav = nav if not config["nav"] else None
        
        if self.nav_config_with_rest:
            # restore explicit config with rest placeholder and build nav
            config["nav"] = self.nav_config_with_rest
            awesome_nav = get_awesome_navigation(files, config)
            
            explicit_nav = get_navigation(files, config)

            for key, value in self.rest_dict.items(): # Add "rest sections" back to nav
                rest_link = Link(AwesomePagesPlugin.REST_PLACEHOLDER, value)
                explicit_nav.items.insert(key, rest_link)
            
        explicit_sections = set(get_by_type(explicit_nav, Section)) if explicit_nav else set()

        if self.nav_config_with_rest:
            self.rest_blocks = self._generate_rest_blocks(awesome_nav.items, [page.file for page in explicit_nav.pages])
            self._insert_rest(explicit_nav.items) 
            nav = explicit_nav

        return AwesomeNavigation(nav.items, Options(**self.config), config["docs_dir"], explicit_sections).to_mkdocs()

    def on_config(self, config: Config):
        for name, plugin in config["plugins"].items():
            if name == "awesome-pages":
                break
            if hasattr(plugin, "on_nav"):
                warnings.warn(NavPluginOrder(name))

        if config["nav"]:
            self._find_rest(config["nav"])
            if self.rest_items:
                self.nav_config_with_rest = config["nav"]

                self.temp_nav_config = config["nav"]

                for index,item in enumerate(config["nav"]):
                    if type(item) is dict: 
                        if AwesomePagesPlugin.REST_PLACEHOLDER in item:
                            self.rest_dict.update({index:item[AwesomePagesPlugin.REST_PLACEHOLDER]})
                            self.temp_nav_config.pop(index)

                config["nav"] = self.temp_nav_config  # Leave non "rest-like" nav, namely for Blog validation
                
        return config
```
```
navigation.py - add imports (needed for func) and new func get_awesome_navigation
...
# Custom get_awesome_navigation
from mkdocs.utils import nest_paths
from mkdocs.structure.files import Files

import logging
log = logging.getLogger(__name__)
from urllib.parse import urlsplit

from typing import TYPE_CHECKING, Any, Iterator, List, Mapping, Optional, Type, TypeVar, Union
from mkdocs.config.defaults import MkDocsConfig
# Custom get_awesome_navigation
...
## Remove config['nav'] from mkdocs.structure.nav > nav_config
def get_awesome_navigation(files: Files, config: Union[MkDocsConfig, Mapping[str, Any]]) -> MkDocsNavigation:
    """Build site navigation from config and files."""
    nav_config = nest_paths(f.src_uri for f in files.documentation_pages())
    items = _data_to_navigation(nav_config, files, config)
    if not isinstance(items, list):
        items = [items]

    pages = _get_by_type(items, Page)

    _add_previous_and_next_links(pages)
    _add_parent_links(items)

    missing_from_config = [file for file in files.documentation_pages() if file.page is None]
    if missing_from_config:
        log.info(
            'The following pages exist in the docs directory, but are not '
            'included in the "nav" configuration:\n  - {}'.format(
                '\n  - '.join(file.src_path for file in missing_from_config)
            )
        )
        for file in missing_from_config:
            Page(None, file, config)

    links = _get_by_type(items, Link)
    for link in links:
        scheme, netloc, path, query, fragment = urlsplit(link.url)
        if scheme or netloc:
            log.debug(f"An external link to '{link.url}' is included in the 'nav' configuration.")
        elif link.url.startswith('/'):
            log.debug(
                f"An absolute path to '{link.url}' is included in the 'nav' "
                "configuration, which presumably points to an external resource."
            )
        else:
            msg = (
                f"A relative path to '{link.url}' is included in the 'nav' "
                "configuration, which is not found in the documentation files"
            )
            log.warning(msg)
    return MkDocsNavigation(items, pages)
```

**Reply by squidfunk** [reply]
Thanks for sharing! As already said, we'll investigate if we can lift this restriction in the future, but for now, other users have something they can build on if they wish to implement the same thing, thanks to you!

**Reply by boozedog** [reply]
@polymorcodeus would you consider submitting a PR with your changes to the [mkdocs-awesome-pages-plugin ](https://github.com/lukasgeiter/mkdocs-awesome-pages-plugin) repo?

**Reply by polymorcodeus** [reply]
@boozedog  Absolutely can - it felt like a bit of a hack and my python is horrible. :) Need to make sure what I actually have rolled and will get it up there.

**Reply by boozedog** [reply]
> @boozedog Absolutely can - it felt like a bit of a hack and my python is horrible. :) Need to make sure what I actually have rolled and will get it up there.

@polymorcodeus Awesome, I did open an issue on that repo as well ... I didn't see a public fork so I could try out your code but I'm happy to test it if you like ... let me know!

**Comment by boozedog** [comment]
so it looks like @polymorcodeus fix won't be integrated into https://github.com/lukasgeiter/mkdocs-awesome-pages-plugin anytime soon (probably not 'til blog becomes a public feature)

@squidfunk would you reconsider providing some default behavior for blog in the nav when nav is not defined? I do like the ability to add pages without manually managing the nav, and my sense is that moar plugins isn't the answer

Any default behavior you provide is fine w/ me, for now I have to disable the blog which makes me sad

**Reply by squidfunk** [reply]
There's an open PR over at Insiders for this right now. I just haven't had the time to review it. It's likely not mergeable as-is, but I will check and reconsider. However, it'll be "just add the blog somewhere", probably at the end. I'm not sure that's good.

**Comment by boozedog** [comment]
@squidfunk at the end is fine with me! I'll shift my commenting to that PR

**Comment by squidfunk** [comment]
https://github.com/squidfunk/mkdocs-material-insiders/pull/64 is an attempt to allow for both, defining and omitting explicit listing blog entries in the navigation. I invite everybody that needs this functionality to test the PR.

**Comment by squidfunk** [comment, marked as answer]
As an update: we're close to finishing a ground-up rewrite of the blog plugin, which removes all of the current limitations and improves interop with third party plugins like awesome pages and friends. I'm sorry this took some time and made so many people upset, but I think the result is a very mature implementation which allows us to take blogging to the next level.

For updates, please follow https://github.com/squidfunk/mkdocs-material/issues/5686

**Comment by StefanoCecere** [comment]
with 9.4.1 it works very well! i just add `- blog` to awesome .pages et voilà. thank you!

**Reply by squidfunk** [reply]
Awesome, glad we solved it!
